File: user-service/src/main/java/user_service/service/AdminService.java
package user_service.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import user_service.dto.RegisterRequest;
import user_service.dto.UpdateUserRoleRequest;
import user_service.dto.UserDto;
import user_service.entity.Role;
import user_service.entity.User;
import user_service.repository.UserRepository;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AdminService {
    
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    
    /**
     * Get all users with pagination
     */
    public Page<UserDto> getAllUsers(Pageable pageable) {
        return userRepository.findAll(pageable)
                .map(this::convertToDto);
    }
    
    /**
     * Get users by role
     */
    public List<UserDto> getUsersByRole(Role role) {
        return userRepository.findByRole(role)
                .stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }
    
    /**
     * Update user role
     */
    @Transactional
    public UserDto updateUserRole(Long userId, UpdateUserRoleRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        user.setRole(request.getRole());
        User updatedUser = userRepository.save(user);
        
        return convertToDto(updatedUser);
    }
    
    /**
     * Delete user
     */
    @Transactional
    public void deleteUser(Long userId) {
        if (!userRepository.existsById(userId)) {
            throw new RuntimeException("User not found");
        }
        userRepository.deleteById(userId);
    }
    
    /**
     * Get user by ID
     */
    public UserDto getUserById(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return convertToDto(user);
    }
    
    /**
     * Get all teachers/instructors
     */
    public List<UserDto> getAllTeachers() {
        return getUsersByRole(Role.INSTRUCTOR);
    }
    
    /**
     * Get all students
     */
    public List<UserDto> getAllStudents() {
        return getUsersByRole(Role.STUDENT);
    }
    
    /**
     * Create instructor account
     */
    @Transactional
    public UserDto createInstructor(RegisterRequest request) {
        // Check if user already exists
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("User with this email already exists");
        }
        
        User instructor = User.builder()
                .username(request.getUsername() != null ? request.getUsername() : request.getEmail())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .role(Role.INSTRUCTOR)
                .createdAt(LocalDateTime.now())
                .build();
        
        User savedInstructor = userRepository.save(instructor);
        return convertToDto(savedInstructor);
    }
    
    /**
     * Get dashboard statistics
     */
    public Map<String, Object> getDashboardStats() {
        Map<String, Object> stats = new HashMap<>();
        
        long totalUsers = userRepository.count();
        long totalStudents = userRepository.countByRole(Role.STUDENT);
        long totalInstructors = userRepository.countByRole(Role.INSTRUCTOR);
        long totalAdmins = userRepository.countByRole(Role.ADMIN);
        
        stats.put("totalUsers", totalUsers);
        stats.put("totalStudents", totalStudents);
        stats.put("totalInstructors", totalInstructors);
        stats.put("totalAdmins", totalAdmins);
        
        return stats;
    }
    
    /**
     * Convert User entity to UserDto
     */
    private UserDto convertToDto(User user) {
        return UserDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .profileImage(user.getProfileImage())
                .role(user.getRole())
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }
}

File: frontend/src/services/adminService.ts
import api from '../utils/api';

export interface User {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
  username: string;
  role: 'STUDENT' | 'INSTRUCTOR' | 'ADMIN';
  isActive: boolean;
  isVerified: boolean;
  createdAt: string;
  lastLoginAt?: string;
}

export interface UpdateUserRoleRequest {
  role: 'STUDENT' | 'INSTRUCTOR' | 'ADMIN';
}

export interface CreateInstructorRequest {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  username?: string;
}

export interface DashboardStats {
  totalUsers: number;
  totalStudents: number;
  totalInstructors: number;
  totalAdmins: number;
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

export const adminService = {
  // Get all users with pagination
  getAllUsers: async (page: number = 0, size: number = 10): Promise<PaginatedResponse<User>> => {
    const response = await api.get<PaginatedResponse<User>>(`/admin/users?page=${page}&size=${size}`);
    return response.data;
  },

  // Get user by ID
  getUserById: async (userId: number): Promise<User> => {
    const response = await api.get<User>(`/admin/users/${userId}`);
    return response.data;
  },
  // Get users by role
  getUsersByRole: async (role: string, page: number = 0, size: number = 10): Promise<PaginatedResponse<User>> => {
    const response = await api.get<PaginatedResponse<User>>(`/admin/users/role/${role}?page=${page}&size=${size}`);
    return response.data;
  },

  // Get all teachers/instructors
  getAllTeachers: async (): Promise<User[]> => {
    const response = await api.get<User[]>('/admin/teachers');
    return response.data;
  },

  // Get all students
  getAllStudents: async (): Promise<User[]> => {
    const response = await api.get<User[]>('/admin/students');
    return response.data;
  },

  // Update user role
  updateUserRole: async (userId: number, request: UpdateUserRoleRequest): Promise<User> => {
    const response = await api.put<User>(`/admin/users/${userId}/role`, request);
    return response.data;
  },

  // Delete user
  deleteUser: async (userId: number): Promise<void> => {
    await api.delete(`/admin/users/${userId}`);
  },

  // Create instructor account
  createInstructor: async (request: CreateInstructorRequest): Promise<User> => {
    const response = await api.post<User>('/admin/instructors', request);
    return response.data;
  },

  // Get dashboard statistics
  getDashboardStats: async (): Promise<DashboardStats> => {
    const response = await api.get<DashboardStats>('/admin/dashboard/stats');
    return response.data;
  }
};
